packages/firebase_performance/lib/src/trace.dart

```diff
@@ -0,0 +1,139 @@
+// Copyright 2017, the Flutter project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+part of firebase_performance;
+
+/// Trace allows you to set beginning and end of a certain action in your app.
+class Trace {
+  /// Maximum allowed length of the Key of the [Trace] attribute.
+  static const int maxAttributeKeyLength = 40;
+
+  /// Maximum allowed length of the Value of the [Trace] attribute.
+  static const int maxAttributeValueLength = 100;
+
+  /// Maximum allowed number of attributes allowed in a trace.
+  static const int maxTraceCustomAttributes = 5;
+
+  /// Maximum allowed length of the name of the [Trace].
+  static const int maxTraceNameLength = 100;
+
+  /// Maximum allowed length of the Key of the [Trace] counter.
+  static const int maxCounterKeyLength = 32;
+
+  final FirebasePerformance _performance;
+
+  /// Id used to sync traces with device platform code.
+  final int _id;
+  bool _hasStarted = false;
+  bool _hasStopped = false;
+
+  /// Name of this [Trace].
+  final String name;
+
+  /// Map of all the counters added to this trace.
+  final HashMap<String, int> counters = new HashMap<String, int>();
+
+  /// Map of all the attributes added to this trace.
+  final HashMap<String, String> attributes = new HashMap<String, String>();
```

Comment:
To match the native SDKs, this should not be mutable. You can call it _attributes and expose an `attributes` getter that provides a read-only view of it. (see [Map.unmodifiable](https://api.dartlang.org/stable/1.24.3/dart-core/Map/Map.unmodifiable.html))